From 0f485b47645ff3a1c453806817667c5a2a387f7c Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 19 Dec 2023 20:17:52 +0100
Subject: [PATCH 0108/1695] plat-stm32mp1: shared_resource: disable MCKPROT if
 not needed

Disable RCC MCKPROT if not needed on STM32MP15 and enforce This change allows
Linux kernel is load and control Cortex-M4 remote processor firmware
when RCC clock and reset controllers are assigned to secure world.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git a03eaf51961ceb1cd5af62d2dc6d3db9868d46ac plat-stm32mp1: shared_resource: disable MCKPROT if not needed]
Change-Id: I18479e734f611033cd481a62f1b66a057e6e5c59
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 .../arch/arm/plat-stm32mp1/shared_resources.c | 25 ++++++++++++-------
 core/include/drivers/stm32mp1_rcc.h           | 11 ++++++++
 2 files changed, 27 insertions(+), 9 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/shared_resources.c b/core/arch/arm/plat-stm32mp1/shared_resources.c
index 76ce64859..d51ec9766 100644
--- a/core/arch/arm/plat-stm32mp1/shared_resources.c
+++ b/core/arch/arm/plat-stm32mp1/shared_resources.c
@@ -663,7 +663,8 @@ static void check_rcc_secure_configuration(void)
 	bool secure = stm32_rcc_is_secure();
 	bool mckprot = stm32_rcc_is_mckprot();
 	enum stm32mp_shres id = STM32MP1_SHRES_COUNT;
-	bool have_error = false;
+	bool need_secure = false;
+	bool need_mckprot = false;
 	uint32_t state = 0;
 
 	if (stm32_bsec_get_state(&state))
@@ -683,21 +684,27 @@ static void check_rcc_secure_configuration(void)
 		    id == STM32MP1_SHRES_SRAM4)
 			continue;
 
-		if ((mckprot_resource(id) && !mckprot) || !secure) {
-			EMSG("RCC %s MCKPROT %s and %s (%u) secure",
-			      secure ? "secure" : "non-secure",
-			      mckprot ? "set" : "not set",
-			      shres2str_id(id), id);
-			have_error = true;
-		}
+		need_secure = true;
+		if (mckprot_resource(id))
+			need_mckprot = true;
+
+		if (!secure || (need_mckprot && !mckprot))
+			EMSG("Error RCC TZEN=%u MCKPROT=%u and %s (%u) secure",
+			      secure, mckprot, shres2str_id(id), id);
 	}
 
-	if (have_error) {
+
+	if ((need_secure && !secure) || (need_mckprot && !mckprot)) {
 		if (IS_ENABLED(CFG_INSECURE))
 			EMSG("WARNING: CFG_INSECURE allows insecure RCC configuration");
 		else
 			panic();
 	}
+
+	if (!need_mckprot && mckprot) {
+		DMSG("Disable RCC MCKPROT");
+		stm32_rcc_set_mckprot(false);
+	}
 }
 
 static void set_gpio_secure_configuration(void)
diff --git a/core/include/drivers/stm32mp1_rcc.h b/core/include/drivers/stm32mp1_rcc.h
index 311cea844..83f3ead80 100644
--- a/core/include/drivers/stm32mp1_rcc.h
+++ b/core/include/drivers/stm32mp1_rcc.h
@@ -561,6 +561,17 @@ static inline bool stm32_rcc_is_mckprot(void)
 {
 	return io_read32(stm32_rcc_base() + RCC_TZCR) & RCC_TZCR_MCKPROT;
 }
+
+static inline void stm32_rcc_set_mckprot(bool enable)
+{
+	vaddr_t tzcr_reg = stm32_rcc_base() + RCC_TZCR;
+
+	if (enable)
+		io_setbits32(tzcr_reg, RCC_TZCR_MCKPROT);
+	else
+		io_clrbits32(tzcr_reg, RCC_TZCR_MCKPROT);
+}
+
 #endif /*__ASSEMBLER__*/
 
 #endif /*__DRIVERS_STM32MP1_RCC_H__*/
-- 
2.47.2

